### PR TITLE
Improvements of the checkpoint functionality and memory occupation of DER

### PR DESCRIPTION
**Pull request by lrzpellegrini**
The main focus of this PR is to introduce various fixes and improvements for the checkpointing functionality:

- The checkpoint functionality now supports a simplified way to register dataset objects so that only constructor parameters are stored
  - This helper is now applied to more datasets
- A de-duplication utility has been added to prevent the in-memory duplication of dataset objects
- Greatly improve the performance of LazyIndices by switching to a NumPy-based storage of indices. This also greatly reduces the time taken to save and load checkpoints
  - Maximum depth before eagerifying reduced to 2
  - Added an additional unit test for LazyRange
- Loading a checkpoint in PyTorch 1.13.* is now handled by a fixed version of `torch.load(...)` (which is bugged in PyTorch 1.13.*)
- Checkpointing has been moved to a separate module to prevent import issues
  - Adapted RTD documentation

Additional elements:
- Fixes to DER memory occupation and checkpoint size:
  - FlatData can now discard (remove the reference to) unused elements (must be enabled via a specific constructor parameter)
    - Useful when storing large amounts of data as dataset data attributes (such as in DER, where logits were saved for each training exemplar)
  - Adapted DER strategy to use this functionality

Misc:
- Added `AvalancheImageNet` dataset, a clone of torchvision ImageNet dataset that supports an additional `meta_root` parameter. Useful when working on certain HPC clusters.

### CODE REVIEW

File: avalanche/benchmarks/utils/data.py [hunk 1]
```diff
@@ -466,9 +466,15 @@ def __init__(
         indices: Optional[List[int]] = None,
         transform_groups: Optional[TransformGroups] = None,
         frozen_transform_groups: Optional[TransformGroups] = None,
+        discard_elements_not_in_indices: bool = False,
```
Comment (AntonioCarta) [inline avalanche/benchmarks/utils/data.py:469]: missing docstring

**Review by AntonioCarta — APPROVED** [review]

**Comment by AntonioCarta** [comment]
Thanks, all of these changes are great. Unfortunately, I messed up the CI by trying to add support for 3.11. Maybe let's wait until I fix it before merging this PR.
